types/src/transaction/mod.rs

```diff
@@ -1105,25 +1106,98 @@ impl TransactionListWithProof {
                 self.transactions.len(),
             );
             itertools::zip_eq(event_lists, self.proof.transaction_infos())
-                .map(|(events, txn_info)| {
-                    let event_hashes: Vec<_> = events.iter().map(ContractEvent::hash).collect();
-                    let event_root_hash =
-                        InMemoryAccumulator::<EventAccumulatorHasher>::from_leaves(&event_hashes)
-                            .root_hash();
-                    ensure!(
-                        event_root_hash == txn_info.event_root_hash(),
-                        "Some event root hash calculated doesn't match that carried on the \
-                         transaction info.",
-                    );
-                    Ok(())
-                })
+                .map(|(events, txn_info)| verify_events_against_root_hash(events, txn_info))
                 .collect::<Result<Vec<_>>>()?;
         }
 
         Ok(())
     }
 }
 
+/// This differs from TransactionListWithProof in that TransactionOutputs are
+/// stored (no transactions). Events are stored inside each TransactionOutput.
+///
+/// Note: the proof cannot verify the TransactionOutputs themselves. This
+/// requires speculative execution of each TransactionOutput to verify that the
+/// resulting state matches the expected state in the proof (for each version).
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct TransactionOutputListWithProof {
+    pub transaction_outputs: Vec<TransactionOutput>,
+    pub first_transaction_output_version: Option<Version>,
+    pub proof: TransactionInfoListProof,
```

Comment:
IMO there's nothing wrong using `TransactionListProof` here. If any, type alias it?

Reply:
See above -- the verify() method is the issue. `TransactionListProof` checks the transactions are tied to the transaction infos, but in `TransactionInfoListProof` we don't ever have access to the transactions.